File: ch8/UVa01099.ml
(* Sharing Chocolate *)
open Scanf
open Printf

let twoPow x = 1 lsl x
let bitOn s x = s land twoPow x <> 0
let setBitOn s x = s lor twoPow x
let getbit s x = (s lsr x) land 1

let n = ref 0
let num_pieces = Array.make 16 0
let area = Array.make 65536 (-1)
let memo = Hashtbl.create 100

let count_area s =
  let ret = ref 0 in
  for i = 0 to !n-1 do
    if bitOn s i then begin
      ret := !ret + num_pieces.(i);
    end
  done;
  !ret

let getOnBits s =
  let rec helper acc i =
    if i < 0 then acc
    else if bitOn s i then helper (i :: acc) (i-1)
    else helper acc (i-1)
  in
  helper [] (!n-1)
;;

let rec dp w s =
  let open Hashtbl in
  if mem memo (w, s) then find memo (w, s)
  else begin
    let a = area.(s) in
    let h = a / w in
    if w = 0 || a mod w <> 0 then begin
      add memo (w, s) false;
      false
    end
    else if mem memo (h, s) then begin
      let ret = find memo (h, s) in
      add memo (w, s) ret;
      ret
    end
    else (
      let bit = Array.of_list (getOnBits s) in
      let l = Array.length bit in
      if l = 1 then begin
        add memo (w, s) true;
        true
      end
      else begin
          let last = twoPow (l-1) - 1 in
          let rec try_using t =
            if t = last then false
            else begin
              let s1 = ref (setBitOn 0 bit.(l-1)) in
              let s2 = ref 0 in
              for k = 0 to l-2 do 
                if bitOn t k then s1 := setBitOn !s1 bit.(k) else s2 := setBitOn !s2 bit.(k)
              done;
              let op1 = dp w !s1 && dp w !s2 in
              let op2 = dp h !s1 && dp h !s2 in
              if op1 || op2 then true else try_using (t+1)
            end
          in
          let ret = try_using 0 in
          add memo (w, s) ret;
          ret
        end
      )
    end

let () =
  let case = ref 1 in
  let rec solve () =
    scanf "%d" (fun k ->
      n := k;
      if !n <> 0 then begin
        scanf "\n%d %d\n" (fun x y ->
          scanf "%d" (fun z -> num_pieces.(0) <- z);
          for i = 1 to !n-1 do scanf " %d" (fun z -> num_pieces.(i) <- z) done;
          scanf "\n" ();
          let target = (twoPow !n) -1 in
          Array.fill area 0 65536 (-1);
          for i = 0 to target do area.(i) <- count_area i done;
          Hashtbl.reset memo;
          let str = if area.(target) <> x * y || not (dp y target) then "No\n" else "Yes\n" in
          printf "Case %d: %s" !case str;
          incr case;
          solve ()
        )
      end
    )
  in 
  solve ()

File: ch8/UVa01099.cpp
// Sharing Chocolate

#include <bits/stdc++.h>
using namespace std;

#define twoPow(X) (1 << (X))
#define bitOn(S, X) (((S) & twoPow(X)) != 0)
#define setBitOn(S, X) ((S) | twoPow(X))
#define getBit(S, X) (((S) >> X) & 1)

int i, n, target, x, y, num_pieces[16], caseNo = 0;
int memo[101][1<<16], area[1<<16];

int count_area(int S) {
  int retval = 0;
  for (int i = 0; i < n; i++)
    if (bitOn(S, i)) // if this bit is on
      retval += num_pieces[i];
  return retval;
}

bool dp(int w, int h, int S) {
  if (memo[w][S] != -1) return memo[w][S];       // note: h is not memo-ed
  int a = area[S];
  if ((w == 0) || (a%w != 0)) return memo[w][S] = false;
  // if (memo[h][S] != -1) return memo[w][S] = memo[h][S];
  vector<short> bit;
  for (int i = 0; i < n; ++i)
    if (bitOn(S, i))
      bit.push_back(i);
  int l = (int)bit.size();
  if (l == 1) // if only one item
    memo[w][S] = true;
  else {
    memo[w][S] = false;
    for (int t = 0; t <= twoPow(l-1)-2; ++t) {
      int s1 = setBitOn(0, bit[l-1]); // Set one always belong to set #1, can reduce search space by half
      int s2 = 0;
      for (int k = 0; k < l-1; ++k)
        if (bitOn(t, k))
          s1 = setBitOn(s1, bit[k]);
        else
          s2 = setBitOn(s2, bit[k]);
      bool op1 = dp(w, area[s1]/w, s1) && dp(w, area[s2]/w, s2); // horizontal cut
      bool op2 = dp(h, area[s1]/h, s1) && dp(h, area[s2]/h, s2); // vertical cut
      if (op1 || op2) {
        memo[w][S] = true;
        break;
      } 
    }     
  }
  return memo[w][S];
}

int main() {
  while (scanf("%d", &n), n) {
    scanf("%d %d", &x, &y);
    for (i = 0; i < n; ++i) scanf("%d", &num_pieces[i]);
    target = twoPow(n)-1;
    memset(area, -1, sizeof area);
    for (i = 0; i <= target; ++i) area[i] = count_area(i); // O(32,768*16 = 524,288)
    memset(memo, -1, sizeof memo);
    printf("Case %d: %s", ++caseNo, (area[target] != x*y || // special case: cannot
                                     !(dp(y, x, target))) ? "No\n" : "Yes\n"); // try subproblems
  }
  return 0;
}
